File: src/main/java/io/vproxy/vfx/ui/stage/VStage.java
package io.vproxy.vfx.ui.stage;

import io.vproxy.base.util.OS;
import io.vproxy.vfx.control.drag.DragHandler;
import io.vproxy.vfx.manager.font.FontManager;
import io.vproxy.vfx.manager.font.FontUsages;
import io.vproxy.vfx.theme.Theme;
import io.vproxy.vfx.ui.scene.VScene;
import io.vproxy.vfx.ui.scene.VSceneGroup;
import io.vproxy.vfx.ui.scene.VSceneGroupInitParams;
import io.vproxy.vfx.ui.scene.VSceneRole;
import io.vproxy.vfx.util.FXUtils;
import javafx.geometry.Insets;
import javafx.scene.Cursor;
import javafx.scene.Scene;
import javafx.scene.control.Label;
import javafx.scene.layout.*;
import javafx.scene.paint.Color;
import javafx.scene.paint.Paint;
import javafx.scene.shape.Circle;
import javafx.stage.Stage;
import javafx.stage.StageStyle;

import java.util.regex.Pattern;

public class VStage {
    public static final int TITLE_BAR_HEIGHT = 28;

    private final Stage stage;
    private final Pane rootContainer = new Pane();
    private final VScene rootContent = new VScene(VSceneRole.MAIN) {{
        enableAutoContentWidthHeight();
    }};
    private final VSceneGroup rootSceneGroup = new VSceneGroup(rootContent,
        new VSceneGroupInitParams()
            .setUseClip(false));
    private final VScene content;
    private final VSceneGroup sceneGroup;
    private final Label title = new Label() {{
        FontManager.get().setFont(FontUsages.windowTitle, this);
        setTextFill(Theme.current().normalTextColor());
        setMouseTransparent(true);
    }};
    private final Circle resizeButton = new Circle() {{
        setStrokeWidth(0);
        setFill(Color.TRANSPARENT);
        setRadius(8);
        setCursor(Cursor.NW_RESIZE);
    }};

    private boolean maximized = false;
    private final MaxResetButton maxrstBtn;

    public VStage() {
        this(new VStageInitParams());
    }

    public VStage(VStageInitParams initParams) {
        this(new Stage(), initParams);
    }

    public VStage(Stage stage) {
        this(stage, new VStageInitParams());
    }

    public VStage(Stage stage, VStageInitParams initParams) {
        this.stage = stage;
        stage.initStyle(StageStyle.TRANSPARENT);

        //noinspection ReplaceNullCheck
        if (initParams.initialScene == null) {
            content = new VScene(VSceneRole.MAIN);
        } else {
            content = initParams.initialScene;
        }
        sceneGroup = new VSceneGroup(content,
            new VSceneGroupInitParams()
                .setGradientCover(true)
                .setUseClip(false));

        var scene = new Scene(rootContainer);
        scene.setFill(Color.TRANSPARENT);
        stage.setScene(scene);
        rootSceneGroup.getNode().setLayoutX(0.5);
        rootSceneGroup.getNode().setLayoutY(0.5);

        FXUtils.makeClipFor(rootSceneGroup.getNode(), 8);

        stage.widthProperty().addListener((ob, old, now) -> {
            if (now == null) return;
            var w = now.doubleValue();
            rootContainer.setPrefWidth(w);
            rootSceneGroup.getNode().setPrefWidth(w - 1);
        });
        stage.heightProperty().addListener((ob, old, now) -> {
            if (now == null) return;
            var h = now.doubleValue();
            rootContainer.setPrefHeight(h);
            rootSceneGroup.getNode().setPrefHeight(h - 1);
        });
        rootContainer.setBackground(Background.EMPTY);
        useDefaultBorder();
        setBackground(Theme.current().sceneBackgroundColor());

        var movingPane = new Pane();
        movingPane.setPrefHeight(TITLE_BAR_HEIGHT);
        var moveWindowHandler = new DragHandler() {
            @Override
            protected void set(double x, double y) {
                if (maximized) {
                    return;
                }
                stage.setX(x);
                stage.setY(y);
            }

            @Override
            protected double[] get() {
                return new double[]{stage.getX(), stage.getY()};
            }
        };
        movingPane.setOnMousePressed(moveWindowHandler);
        movingPane.setOnMouseDragged(moveWindowHandler);
        rootContent.getContentPane().getChildren().add(movingPane);

        var closeBtn = new CloseButton(this, initParams);
        if (initParams.closeButton) rootContent.getContentPane().getChildren().add(closeBtn);

        maxrstBtn = new MaxResetButton(this, initParams);
        if (initParams.maximizeAndResetButton) rootContent.getContentPane().getChildren().add(maxrstBtn);

        var iconifyBtn = new IconifyButton(this, initParams);
        if (initParams.iconifyButton) rootContent.getContentPane().getChildren().add(iconifyBtn);

        sceneGroup.getNode().setLayoutX(0.5);
        sceneGroup.getNode().setLayoutY(TITLE_BAR_HEIGHT);
        rootContent.getContentPane().getChildren().add(sceneGroup.getNode());

        rootContent.getContentPane().getChildren().add(title);

        var resizeHandler = new DragHandler() {
            @Override
            protected void set(double x, double y) {
                if (x < 256) {
                    x = 256;
                }
                if (y < 128) {
                    y = 128;
                }
                stage.setWidth(x);
                stage.setHeight(y);
            }

            @Override
            protected double[] get() {
                return new double[]{
                    stage.getWidth(),
                    stage.getHeight(),
                };
            }
        };
        resizeButton.setOnMousePressed(resizeHandler);
        resizeButton.setOnMouseDragged(resizeHandler);

        title.textProperty().addListener((ob, old, now) -> updateTitlePosition());
        stage.widthProperty().addListener((ob, old, now) -> {
            if (now == null) return;
            var w = now.doubleValue();
            final double padRight = 0;
            var controlButtonsOffset = w - padRight;
            if (initParams.closeButton) {
                controlButtonsOffset -= CloseButton.WIDTH;
                closeBtn.setLayoutX(controlButtonsOffset);
            }
            if (initParams.maximizeAndResetButton) {
                controlButtonsOffset -= MaxResetButton.WIDTH;
                maxrstBtn.setLayoutX(controlButtonsOffset);
            }
            if (initParams.iconifyButton) {
                controlButtonsOffset -= IconifyButton.WIDTH;
                iconifyBtn.setLayoutX(controlButtonsOffset);
            }
            movingPane.setPrefWidth(w);
            rootSceneGroup.getNode().setPrefWidth(w - 1);
            resizeButton.setLayoutX(w - 8);
            updateTitlePosition();
        });
        stage.heightProperty().addListener((ob, old, now) -> {
            if (now == null) return;
            var h = now.doubleValue();
            rootSceneGroup.getNode().setPrefHeight(h - 1);
            resizeButton.setLayoutY(h - 8);
        });
        FXUtils.observeWidthHeight(rootContent.getContentPane(), sceneGroup.getNode(), 0, -TITLE_BAR_HEIGHT);

        rootContainer.getChildren().add(rootSceneGroup.getNode());
        if (initParams.resizable) {
            rootContainer.getChildren().add(resizeButton);
        }

        stage.setOnCloseRequest(e -> close());
    }

    public void useDefaultBorder() {
        useLightBorder();
    }

    public void useLightBorder() {
        rootContainer.setBorder(new Border(new BorderStroke(
            Theme.current().windowBorderColorLight(),
            BorderStrokeStyle.SOLID,
            new CornerRadii(8),
            new BorderWidths(0.5))));
    }

    public void useDarkBorder() {
        rootContainer.setBorder(new Border(new BorderStroke(
            Theme.current().windowBorderColorDark(),
            BorderStrokeStyle.SOLID,
            new CornerRadii(8),
            new BorderWidths(0.5))));
    }

    private void updateTitlePosition() {
        if (title.getText() == null || title.getText().isBlank()) {
            title.setVisible(false);
            return;
        }
        title.setVisible(true);
        var bounds = FXUtils.calculateTextBounds(title);
        var tWidth = bounds.getWidth();
        var tHeight = bounds.getHeight();
        var sWidth = stage.getWidth();
        title.setLayoutX((sWidth - tWidth) / 2);
        title.setLayoutY((TITLE_BAR_HEIGHT - tHeight) / 2);
    }

    public void setBackground(Paint paint) {
        rootSceneGroup.getNode().setBackground(new Background(new BackgroundFill(
            paint,
            CornerRadii.EMPTY,
            Insets.EMPTY)));
    }

    public Stage getStage() {
        return stage;
    }

    public VSceneGroup getRootSceneGroup() {
        return rootSceneGroup;
    }

    public VScene getRoot() {
        return rootContent;
    }

    public VSceneGroup getSceneGroup() {
        return sceneGroup;
    }

    public VScene getInitialScene() {
        return content;
    }

    public void setTitle(String title) {
        this.title.setText(title);
        stage.setTitle(title);
    }

    private boolean supportsIconify() {
        if (!OS.isMac()) {
            return true;
        }
        var pattern = Pattern.compile("^(\\d+)(.*)$");
        var version = System.getProperty("javafx.runtime.version", "");
        var matcher = pattern.matcher(version);
        if (!matcher.find() || matcher.groupCount() < 1) {
            return false;
        }
        int v;
        try {
            v = Integer.parseInt(matcher.group(1));
        } catch (NumberFormatException e) {
            return false;
        }
        return v >= 17;
    }

    public void setIconified(boolean iconified) {
        if (supportsIconify()) {
            stage.setIconified(iconified);
        } else {
            if (iconified) {
                stage.toBack();
            }
        }
    }

    private double stageOriginalX;
    private double stageOriginalY;
    private double stageOriginalW;
    private double stageOriginalH;

    public void setMaximized(boolean maximized) {
        if (this.maximized == maximized) {
            return;
        }
        this.maximized = maximized;
        maxrstBtn.updateImage();

        if (OS.isMac()) {
            if (maximized) {
                stageOriginalX = stage.getX();
                stageOriginalY = stage.getY();
                stageOriginalW = stage.getWidth();
                stageOriginalH = stage.getHeight();
                var screen = FXUtils.getScreenOf(stage);
                if (screen == null) {
                    return;
                }
                var bounds = screen.getBounds();
                stage.setX(bounds.getMinX());
                stage.setY(bounds.getMinY() + 24);
                stage.setWidth(bounds.getWidth());
                stage.setHeight(bounds.getHeight() - 24);
            } else {
                stage.setX(stageOriginalX);
                stage.setY(stageOriginalY);
                stage.setWidth(stageOriginalW);
                stage.setHeight(stageOriginalH);
            }
        } else {
            stage.setMaximized(maximized);
        }
    }

    public boolean isMaximized() {
        return maximized;
    }

    public void close() {
        stage.close();
    }

    public void show() {
        stage.show();
    }

    public void showAndWait() {
        stage.showAndWait();
    }

    public void temporaryOnTop() {
        if (stage.isAlwaysOnTop()) {
            return;
        }
        stage.setAlwaysOnTop(true);
        FXUtils.runDelay(500, () -> stage.setAlwaysOnTop(false));
    }
}

File: src-ts/src/ui/stage/VStage.ts
import { Pane } from '../../javafx/Pane.js';
import { Label } from '../../javafx/Label.js';
import { Stage } from '../../javafx/stage.js';
import { Scene } from '../../javafx/scene.js';
import { Color, Paint } from '../../javafx/color.js';
import { Circle } from '../../javafx/shapes.js';
import {
  Background,
  BackgroundFill,
  Border,
  BorderStroke,
  BorderStrokeDescriptor,
  BorderStrokeStyle,
  CornerRadii,
  BorderWidths,
  Insets,
} from '../../javafx/layout.js';
import { FXUtils } from '../../util/FXUtils.js';
import { Theme } from '../../theme/Theme.js';
import { FontManager } from '../../manager/font/FontManager.js';
import { FontUsages } from '../../manager/font/FontUsages.js';
import { VScene } from '../scene/VScene.js';
import { VSceneGroup } from '../scene/VSceneGroup.js';
import { VSceneGroupInitParams } from '../scene/VSceneGroupInitParams.js';
import { VSceneRole } from '../scene/VSceneRole.js';
import { VStageInitParams } from './VStageInitParams.js';
import { CloseButton } from './CloseButton.js';
import { MaxResetButton } from './MaxResetButton.js';
import { IconifyButton } from './IconifyButton.js';
import { OS } from '../../vproxy-base/OS.js';

export class VStage {
  static readonly TITLE_BAR_HEIGHT = 28;
  // Floor for manual resize. The OS window is non-resizable so it won't clamp
  // size itself; the resize handle enforces these instead. Mirrors the
  // minWidth/minHeight declared for the main window in tauri.conf.json.
  static readonly MIN_WIDTH = 256;
  static readonly MIN_HEIGHT = 128;

  private readonly stage: Stage;
  private readonly rootContainer: Pane = new Pane();
  private readonly rootContent: VScene;
  private readonly rootSceneGroup: VSceneGroup;
  private readonly content: VScene;
  private readonly sceneGroup: VSceneGroup;
  private readonly title: Label;
  private readonly resizeButton: Circle;

  private maximized = false;
  private readonly maxrstBtn: MaxResetButton;

  constructor();
  constructor(initParams: VStageInitParams);
  constructor(stage: Stage);
  constructor(stage: Stage, initParams: VStageInitParams);
  constructor(stageOrInit?: Stage | VStageInitParams, maybeInit?: VStageInitParams) {
    let stage: Stage;
    let initParams: VStageInitParams;
    if (stageOrInit instanceof Stage) {
      stage = stageOrInit;
      initParams = maybeInit ?? new VStageInitParams();
    } else if (stageOrInit instanceof VStageInitParams) {
      stage = new Stage();
      initParams = stageOrInit;
    } else {
      stage = new Stage();
      initParams = new VStageInitParams();
    }
    this.stage = stage;
    stage.initStyle('TRANSPARENT');

    this.rootContent = new VScene(VSceneRole.MAIN);
    this.rootContent.enableAutoContentWidthHeight();

    this.rootSceneGroup = new VSceneGroup(
      this.rootContent,
      new VSceneGroupInitParams().setUseClip(false),
    );

    if (initParams.initialScene === null) {
      this.content = new VScene(VSceneRole.MAIN);
    } else {
      this.content = initParams.initialScene;
    }
    this.sceneGroup = new VSceneGroup(
      this.content,
      new VSceneGroupInitParams().setGradientCover(true).setUseClip(false),
    );

    const scene = new Scene(this.rootContainer);
    scene.setFill(Color.TRANSPARENT);
    this.rootSceneGroup.getNode().setLayoutX(0.5);
    this.rootSceneGroup.getNode().setLayoutY(0.5);

    FXUtils.makeRoundedClipFor(this.rootSceneGroup.getNode() as Pane, 8);

    stage.widthProperty.addListener(() => {
      const w = stage.getWidth();
      this.rootContainer.setPrefWidth(w);
      (this.rootSceneGroup.getNode() as Pane).setPrefWidth(w - 1);
    });
    stage.heightProperty.addListener(() => {
      const h = stage.getHeight();
      this.rootContainer.setPrefHeight(h);
      (this.rootSceneGroup.getNode() as Pane).setPrefHeight(h - 1);
    });
    this.rootContainer.setBackground(Background.EMPTY);
    this.useDefaultBorder();
    this.setBackground(Theme.current().sceneBackgroundColor());

    const movingPane = new Pane();
    movingPane.setPrefHeight(VStage.TITLE_BAR_HEIGHT);
    movingPane.el.style.position = 'absolute';
    movingPane.el.style.top = '0';
    movingPane.el.style.left = '0';
    movingPane.el.setAttribute('data-tauri-drag-region', '');
    this.rootContent.getContentPane().getChildren().add(movingPane);

    const closeBtn = new CloseButton(this, initParams);
    if (initParams.closeButton) this.rootContent.getContentPane().getChildren().add(closeBtn);

    this.maxrstBtn = new MaxResetButton(this, initParams);
    if (initParams.maximizeAndResetButton) this.rootContent.getContentPane().getChildren().add(this.maxrstBtn);

    const iconifyBtn = new IconifyButton(this, initParams);
    if (initParams.iconifyButton) this.rootContent.getContentPane().getChildren().add(iconifyBtn);

    this.sceneGroup.getNode().setLayoutX(0.5);
    this.sceneGroup.getNode().setLayoutY(VStage.TITLE_BAR_HEIGHT);
    this.rootContent.getContentPane().getChildren().add(this.sceneGroup.getNode());

    this.title = new Label();
    FontManager.get().setFont(FontUsages.windowTitle, this.title);
    this.title.setTextFill(Theme.current().normalTextColor());
    this.title.setMouseTransparent(true);
    this.rootContent.getContentPane().getChildren().add(this.title);

    // Right-bottom resize handle — the ONLY way to resize. The OS window is
    // created non-resizable (resizable:false in tauri.conf.json /
    // WindowManager.open) so the native borders/corners can't be dragged.
    //
    // The catch: setSize() is a no-op on a non-resizable window (the OS rejects
    // all size changes for that window style). So we can't just keep it false
    // and call setSize. startResizeDragging() is likewise a no-op. The chosen
    // approach: flip the window to resizable:true for the duration of the drag
    // (so setSize works), then flip it back to false when the gesture ends,
    // restoring the "no edge resize" state.
    this.resizeButton = new Circle();
    this.resizeButton.setStrokeWidth(0);
    this.resizeButton.setFill(Color.TRANSPARENT);
    this.resizeButton.setRadius(8);
    this.resizeButton.setCursor('nwse-resize');
    this.resizeButton.el.style.position = 'absolute';
    this.resizeButton.el.style.touchAction = 'none';
    this.resizeButton.el.addEventListener('pointerdown', (e) => {
      e.preventDefault();
      e.stopPropagation();
      const win = stage.tauriWindow;
      const startW = stage.getWidth();
      const startH = stage.getHeight();
      const startX = e.clientX;
      const startY = e.clientY;
      this.resizeButton.el.setPointerCapture(e.pointerId);

      // Enable resizing just for this gesture. Fire-and-forget the await: the
      // first few pointermove events may land before the IPC resolves, in which
      // case their setSize calls are coalesced/dropped harmlessly by Stage; the
      // gesture spans many events so resize kicks in within a frame or two.
      // Requires the core:window:allow-set-resizable capability.
      const enable = win.setResizable(true).catch((err) => {
        console.error('[vfx:resize] setResizable(true) failed', err);
      });

      // Freeze the layout cascade for the duration of the drag: the OS window
      // resizes every frame, but width/heightProperty suppress dispatch so the
      // observe-width/height reflow chain (and all the scene/table/scroll
      // relayout it triggers) stays paused. Content appears frozen while the
      // bare window outline follows the cursor. `finish` un-pauses and the UI
      // reflows exactly once to the final size.
      stage.setResizing(true);

      const onMove = (ev: PointerEvent): void => {
        let w = startW + (ev.clientX - startX);
        let h = startH + (ev.clientY - startY);
        if (w < VStage.MIN_WIDTH) w = VStage.MIN_WIDTH;
        if (h < VStage.MIN_HEIGHT) h = VStage.MIN_HEIGHT;
        void stage.setWidth(w);
        void stage.setHeight(h);
      };
      const finish = (ev: PointerEvent): void => {
        this.resizeButton.el.releasePointerCapture(ev.pointerId);
        this.resizeButton.el.removeEventListener('pointermove', onMove);
        this.resizeButton.el.removeEventListener('pointerup', finish);
        this.resizeButton.el.removeEventListener('pointercancel', finish);
        // Un-freeze first: flush the final size to listeners so content snaps
        // to the new dimensions in one pass, THEN re-lock the borders.
        stage.setResizing(false);
        void enable.then(() => {
          win.setResizable(false).catch((err) => {
            console.error('[vfx:resize] setResizable(false) failed', err);
          });
        });
      };
      this.resizeButton.el.addEventListener('pointermove', onMove);
      this.resizeButton.el.addEventListener('pointerup', finish);
      this.resizeButton.el.addEventListener('pointercancel', finish);
    });

    this.title.textProperty.addListener(() => this.updateTitlePosition());
    stage.widthProperty.addListener(() => {
      const w = stage.getWidth();
      const padRight = 0;
      let controlButtonsOffset = w - padRight;
      if (initParams.closeButton) {
        controlButtonsOffset -= CloseButton.WIDTH;
        closeBtn.setLayoutX(controlButtonsOffset);
      }
      if (initParams.maximizeAndResetButton) {
        controlButtonsOffset -= MaxResetButton.WIDTH;
        this.maxrstBtn.setLayoutX(controlButtonsOffset);
      }
      if (initParams.iconifyButton) {
        controlButtonsOffset -= IconifyButton.WIDTH;
        iconifyBtn.setLayoutX(controlButtonsOffset);
      }
      movingPane.setPrefWidth(w);
      (this.rootSceneGroup.getNode() as Pane).setPrefWidth(w - 1);
      this.resizeButton.setCenterX(w - 8);
      this.updateTitlePosition();
    });
    stage.heightProperty.addListener(() => {
      const h = stage.getHeight();
      (this.rootSceneGroup.getNode() as Pane).setPrefHeight(h - 1);
      this.resizeButton.setCenterY(h - 8);
    });
    FXUtils.observeWidthHeight(
      this.rootContent.getContentPane(),
      this.sceneGroup.getNode() as Pane,
      0,
      -VStage.TITLE_BAR_HEIGHT,
    );

    this.rootContainer.getChildren().add(this.rootSceneGroup.getNode());
    if (initParams.resizable) {
      this.rootContainer.getChildren().add(this.resizeButton);
    }

    stage.onCloseRequestedHandlers.add(() => { void this.close(); });
  }

  useDefaultBorder(): void { this.useLightBorder(); }

  useLightBorder(): void {
    this.rootContainer.setBorder(new Border(new BorderStroke(new BorderStrokeDescriptor(
      Theme.current().windowBorderColorLight(),
      BorderStrokeStyle.SOLID,
      CornerRadii.uniform(8),
      BorderWidths.uniform(0.5),
    ))));
  }

  useDarkBorder(): void {
    this.rootContainer.setBorder(new Border(new BorderStroke(new BorderStrokeDescriptor(
      Theme.current().windowBorderColorDark(),
      BorderStrokeStyle.SOLID,
      CornerRadii.uniform(8),
      BorderWidths.uniform(0.5),
    ))));
  }

  private updateTitlePosition(): void {
    const text = this.title.getText();
    if (text === null || text.trim() === '') {
      this.title.setVisible(false);
      return;
    }
    this.title.setVisible(true);
    const bounds = FXUtils.calculateTextBounds(this.title);
    const tWidth = bounds.width;
    const tHeight = bounds.height;
    const sWidth = this.stage.getWidth();
    this.title.setLayoutX((sWidth - tWidth) / 2);
    this.title.setLayoutY((VStage.TITLE_BAR_HEIGHT - tHeight) / 2);
  }

  setBackground(paint: Paint): void {
    (this.rootSceneGroup.getNode() as Pane).setBackground(new Background(new BackgroundFill(
      paint,
      CornerRadii.EMPTY,
      Insets.EMPTY,
    )));
  }

  getStage(): Stage { return this.stage; }
  getRootSceneGroup(): VSceneGroup { return this.rootSceneGroup; }
  getRoot(): VScene { return this.rootContent; }
  getSceneGroup(): VSceneGroup { return this.sceneGroup; }
  getInitialScene(): VScene { return this.content; }

  setTitle(t: string): void {
    this.title.setText(t);
    void this.stage.setTitle(t);
  }

  setIconified(iconified: boolean): void {
    void this.stage.setIconified(iconified);
  }

  private stageOriginalX = 0;
  private stageOriginalY = 0;
  private stageOriginalW = 0;
  private stageOriginalH = 0;

  async setMaximized(maximized: boolean): Promise<void> {
    if (this.maximized === maximized) return;
    this.maximized = maximized;
    this.maxrstBtn.updateImage();

    if (OS.isMacSync()) {
      if (maximized) {
        this.stageOriginalX = this.stage.getX();
        this.stageOriginalY = this.stage.getY();
        this.stageOriginalW = this.stage.getWidth();
        this.stageOriginalH = this.stage.getHeight();
        // Tauri has no direct screen-bounds API exposed; fall through
        // to the generic maximize below.
      } else {
        await this.stage.setX(this.stageOriginalX);
        await this.stage.setY(this.stageOriginalY);
        await this.stage.setWidth(this.stageOriginalW);
        await this.stage.setHeight(this.stageOriginalH);
        return;
      }
    }
    await this.stage.setMaximized(maximized);
  }

  isMaximized(): boolean { return this.maximized; }

  close(): Promise<void> { return this.stage.close(); }

  show(): Promise<void> { return this.stage.show(); }
  showAndWait(): Promise<void> { return this.stage.showAndWait(); }

  async temporaryOnTop(): Promise<void> {
    const always = await this.stage.tauriWindow.isAlwaysOnTop();
    if (always) return;
    await this.stage.setAlwaysOnTop(true);
    FXUtils.runDelay(500, () => { void this.stage.setAlwaysOnTop(false); });
  }
}
